Initial commit for the NVMeoF gateway to Ceph management daemon (#3)
ceph/ceph-nvmeof: README.md

# nvmeof-gateway

Management gateway daemon to setup access to Ceph storage over NVMeoF 

This daemon runs as root. It provides the ability to export existing RBD images as NVMeoF namespaces. Creation of RBD images is not within the scope of this daemon.

# Initial configuration

1. The daemon is a gRPC server, so the host running the server will need to install gRPC packages:

		$ python3 -m pip install grpcio
		$ python3 -m pip install grpcio-tools
		
2. Modify the config file (default nvme_gw.config) to reflect the IP/ Port where the server can be reached:

		gateway_addr = <IP address at which the client can reach the gateway>
		gateway_port = <port at which the client can reach the gateway>
	
3. To [enable mTLS](#mtls-configuration-for-testing-purposes) using self signed certificates, edit the config file to set:

		enable_auth = True  # Setting this to False will open an insecure port		

> 4. Compiling protobuf files for gRPC:

Reviewer: Ilya suggested to change this to `Compile protobuf files for gRPC`.

		$ sh -c 'echo 4096 > /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages'

7. Start the gateway server daemon:
		
		$ python3 nvme_gw_server.py [-c config_filename]

# CLI Usage

The CLI tool can be used to initiate a connection to the gateway and run commands to configure the NVMe targets. 

Run the tool with the -h flag to see a list of available commands:
	
	$ python3 ./nvme_gw_cli.py -h
	usage: python3 ./nvme_gw_cli.py [-h] [-c CONFIG]
			{create_bdev,create_subsystem,create_namespace,allow_any_hosts,create_transport,create_listener} ...

	CLI to manage NVMe gateways

	positional arguments:
	{create_bdev,create_subsystem,create_namespace,allow_any_hosts,create_transport,create_listener}

	optional arguments:
	-h, --help            			show this help message and exit
	-c CONFIG, --config CONFIG
			      			Path to config file

Example:

	$ python3 ./nvme_gw_cli.py create_bdev -h
	usage: python3 ./nvme_gw_cli.py create_bdev [-h] -i IMAGE -p POOL [-b BDEV_NAME] [-u USER] [-s BLOCK_SIZE]

	optional arguments:
	-h, --help            			show this help message and exit
	-i IMAGE, --image IMAGE
			      			RBD image name
	-p POOL, --pool POOL  			Ceph pool name

> 	-b BDEV_NAME, --bdev_name BDEV_NAME

Reviewer: ```suggestion
	-b BDEV_NAME, --bdev BDEV_NAME
```

# nvmeof-gateway

Management gateway daemon to setup access to Ceph storage over NVMeoF 

This daemon runs as root. It provides the ability to export existing RBD images as NVMeoF namespaces. Creation of RBD images is not within the scope of this daemon.

> # Initial configuration:

Reviewer 2: ```suggestion
# Initial configuration
```

Reviewer: Note, in Ilya's suggested change it is without the colon symbol.

1. The daemon is a gRPC server, so the host running the server will need to install gRPC packages:

		$ python3 -m pip install grpcio
		$ python3 -m pip install grpcio-tools
		
2. Modify the config file (default nvme_gw.config) to reflect the IP/ Port where the server can be reached:

		gateway_addr = <IP address at which the client can reach the gateway>
		gateway_port = <port at which the client can reach the gateway>
	

> 2. To [enable mTLS](#mtls-configuration-for-testing-purposes) using self signed certificates, edit the config file to set:

Reviewer 2: Nit: numbering is off here.  I know it doesn't matter for rendering but some people use plain text editors ;)

		enable_auth = True  # Setting this to False will open an insecure port		

> 3. Compiling protobuf files for grpc:

Reviewer 2: ```suggestion
3. Compile protobuf files for gRPC:
```

		
	    $ python3 -m grpc_tools.protoc --proto_path=./proto ./proto/nvme_gw.proto --python_out=. --grpc_python_out=.   

4. SPDK v21.04 is included in this repository. Edit the config file to set:

		spdk_path = <complete path to SPDK parent directory>
		spdk_tgt = <relative path to SPDK target executable>

> 5. Setup SPDK

Reviewer 2: add an empty line above

	Navigate to the spdk folder & install dependencies:

		$ ./scripts/pkgdep.sh

> 	Intialize configuration:

Reviewer 2: ```suggestion
	Initialize configuration:
```

		$ apt install librbd-dev
		$ ./configure --with-rbd

	Build the SPDK app:

	    $ make

	SPDK requires hugepages to be set up:

		$ sh -c 'echo 4096 > /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages'

6. Start the gateway server daemon:
		
		$ python3 nvme_gw_server.py [-c config_filename]

# CLI Usage

> The CLI code can be used to initiate a connection to the gateway and run commands to configure the NVMe targets. 

Reviewer 2: ```suggestion
The CLI tool can be used to initiate a connection to the gateway and run commands to configure the NVMe targets. 
```

> Run the code with the -h flag to see a list of available commands:

Reviewer 2: ```suggestion
Run the tool with the -h flag to see a list of available commands:
```

	
	$ python3 ./nvme_gw_cli.py -h
	usage: python3 ./nvme_gw_cli.py [-h] [-c CONFIG]
			{create_bdev,create_subsystem,create_namespace,allow_any_hosts,create_transport,create_listener} ...

	CLI to manage NVMe gateways

	positional arguments:
	{create_bdev,create_subsystem,create_namespace,allow_any_hosts,create_transport,create_listener}

	optional arguments:
	-h, --help            			show this help message and exit
	-c CONFIG, --config CONFIG
			      			Path to config file

Example:

	$ python3 ./nvme_gw_cli.py create_bdev -h
	usage: python3 ./nvme_gw_cli.py create_bdev [-h] -i IMAGE -p POOL [-b BDEV_NAME] [-u USER] [-s BLOCK_SIZE]

	optional arguments:
	-h, --help            			show this help message and exit
	-i IMAGE, --image IMAGE
			      			RBD image name
	-p POOL, --pool POOL  			Ceph pool name
	-b BDEV_NAME, --bdev_name BDEV_NAME
						Bdev name
	-u USER, --user USER  			User ID
	-s BLOCK_SIZE, --block_size BLOCK_SIZE
						Block size

# mTLS Configuration for testing purposes

> For testing purposes, self signed certificates & keys can be generated locally using OpenSSL.

Reviewer 2: don't use `&` in English text
```suggestion
For testing purposes, self signed certificates and keys can be generated locally using OpenSSL.
```

> For the server, generate credentials for server name 'my.server' in files called server.key & server.crt:

Reviewer 2: ```suggestion
For the server, generate credentials for server name 'my.server' in files called server.key and server.crt:
```

>   	% openssl req -x509 -newkey rsa:4096 -nodes -keyout server.key -out server.crt -days 3650 -subj '/CN=my.server'

Reviewer 2: stick to `$` for denoting the prompt
```suggestion
  	$ openssl req -x509 -newkey rsa:4096 -nodes -keyout server.key -out server.crt -days 3650 -subj '/CN=my.server'
```

For client:

>   	% openssl req -x509 -newkey rsa:4096 -nodes -keyout client.key -out client.crt -days 3650 -subj '/CN=client1'

Reviewer 2: ```suggestion
  	$ openssl req -x509 -newkey rsa:4096 -nodes -keyout client.key -out client.crt -days 3650 -subj '/CN=client1'
```

Indicate the location of the keys and certificates in the config file:

	[mtls]

	server_key = ./server.key
	client_key = ./client.key
	server_cert = ./server.crt
	client_cert = ./client.crt

# Example NVMe volume access

1. Start the gateway server:

		$ python3 nvme_gw_server.py
		INFO:root:SPDK PATH: /path/to/spdk

> 		INFO:root:Starting /path.to/spdk/tgt/nvmf_tgt all -u

Reviewer 2: ```suggestion
		INFO:root:Starting /path/to/spdk/tgt/nvmf_tgt all -u
```

# nvmeof-gateway

> Management gateway daemon to setup access to Ceph storage over NVMeoF 

Reviewer 2: Nit: add an empty line before this line

Initial configuration:

> 1. This daemon runs as root. User may provide the path to the gateway configuration file (default ./nvme_gateway.config) as an argument.

Reviewer 2: ```suggestion
1. This daemon runs as root. User may provide the path to the gateway configuration file (default ./nvme_gw.config) as an argument.
```
here and elsewhere -- this is what it actually defaults to

Reviewer 2: Or maybe the intent was to default to `nvme_gateway.config` for the server and `nvme_client.config` for the CLI?

		python3 nvme_gw_server.py -c config_filename

2. To enable mTLS using self signed certificates, edit the nvme_gateway.config file to set:

		enable_auth = True  # Setting this to False will open an insecure port
		gateway_addr = <IP address at which the client can reach the gateway>
		gateway_port = <port at which the client can reach the gateway>

3. Compiling protobuf files for grpc:

		python3 -m grpc_tools.protoc --proto_path=./proto ./proto/nvme_gw.proto --python_out=. --grpc_python_out=.   

4. RBD images needed to be exported as NVMe namespaces must exist

5. SPDK v21.04 is included in this repository. Edit the nvme_gateway.config file to set:

		spdk_path = <complete path to SPDK parent directory>

> 		spdk_tgt = <complete path to SPDK target executable>

Reviewer 2: AFAIT `spdk_tgt` is actually a relative path -- it is joined with `spdk_path`.

Author: true.

# SPDK Setup

cd to the spdk folder

Install dependencies:

	./scripts/pkgdep.sh

Intialize configuration:

	./configure --with-rbd

Build the SPDK app:

        make

Before running the SPDK target, setup huge pages:

	sudo sh -c 'echo 4096 > /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages'

# mTLS Configuration for testing purposes

> mTLS can be enabled by setting enable_auth=True in both the client & server scripts

Reviewer 2: ```suggestion
mTLS can be enabled by setting enable_auth=True in both the client & server configuration files.
```

For testing purposes, self signed ertificates & keys can be generated locally using openssl:

For the server, generate credentials for server name 'my.server' in files called server.key & server.crt

  	% openssl req -x509 -newkey rsa:4096 -nodes -keyout server.key -out server.crt -days 3650 -subj '/CN=my.server'

For client:

  	% openssl req -x509 -newkey rsa:4096 -nodes -keyout client.key -out client.crt -days 3650 -subj '/CN=client1'

> Indicate the location of the keys and certificates in the nvme_gateway.conf file under the [mtls] section

Reviewer 2: ```suggestion
Indicate the location of the keys and certificates in the nvme_gw.config file under the [mtls] section
```

# CLI Usage

The CLI code can be used to initiate a connection to the gateway and run commands to configure the NVMe targets. Run the code with the -h flag to see a list of available commands:
	
	$ python3 ./nvme_gw_cli.py -h

> 	usage: nvme_gw_cli [-h] [-c config_FILE]

Reviewer 2: Need to pick a case here and below ;)

Author: ok, if you insist! :)

                   {create_bdev,create_subsystem,create_namespace,allow_any_hosts,create_transport,create_listener} ...

	CLI to manage NVMe gateways

	positional arguments:
  		{create_bdev,create_subsystem,create_namespace,allow_any_hosts,create_transport,create_listener}

	optional arguments:
  		-h, --help            show this help message and exit
  		-c config_FILE, --config config_FILE
                        Path to config file

Specify a command with the -h flag to see more information about the command, including a list of optional and required parameters:

	$ python3 ./nvme_gw_cli.py create_bdev -h
	usage: nvme_gw_cli create_bdev [-h] -r RBD -p POOL [-n NAME] [-u USER] [-b BLOCK_SIZE]

	Creates a bdev from a Ceph RBD.

	optional arguments:
  	-h, --help            show this help message and exit
  	-r RBD, --rbd RBD     RBD image name
  	-p POOL, --pool POOL  Ceph pool name
	-n NAME, --name NAME  Bdev name
	-u USER, --user USER  User ID
	-b BLOCK_SIZE, --block_size BLOCK_SIZE

> 				Block size

Reviewer 2: Would be nice to have a complete example of setting up a namespace and mounting e.g. an ext4 filesystem on the initiator side.

# nvmeof-gateway
Management gateway daemon to setup access to Ceph storage over NVMeoF 

Initial configuration:

1. This daemon runs as root. User must provide the path to the gateway configuration file (nvme_gateway.config) as an argument.

2. To enable mTLS using self signed certificates, edit the nvme_gateway.config file to set:

		enable_auth = True  # Setting this to False will open an insecure port
		gateway_host = <IP address at which the client can reach the gateway>
		gateway_port = <port at which the client can reach the gateway>

3. Compiling protobuf files for grpc:

		python3 -m grpc_tools.protoc --proto_path=./proto ./proto/nvme_gw.proto --python_out=. --grpc_python_out=.   

4. RBD images needed to be exported as NVMe namespaces must exist

5. SPDK v21.04 must be [checked out and set up](https://github.ibm.com/genctl-acadia/nvmeof-gateway#SPDK-Setup). Edit the nvme_gateway.config file to set:

		spdk_path = <path to SPDK parent directory>
		spdk_tgt = <path to SPDK target executable>

# SPDK Setup

	apt install librbd-dev

	git clone https://github.com/spdk/spdk.git
	git checkout v21.04

> 	git submodule update --init

Reviewer: As the ceph-nvmeof already includes spdk as a submodule, may be it makes sense to update to v21.04 and use it in the instructions?

Author: agreed. this is an artifact of how we tested in our local setup. This repo already has v21.04 included.

  	% openssl req -x509 -newkey rsa:4096 -nodes -keyout server.key -out server.crt -days 3650 -subj '/CN=my.server'

For client:

  	% openssl req -x509 -newkey rsa:4096 -nodes -keyout client.key -out client.crt -days 3650 -subj '/CN=client1'

Indicate the loaction of the keys and certificates in the nvme_gateway.conf file under the [MTLS] section

# CLI Usage

The CLI code can be used to initiate a connection to the gateway and run commands to configure the NVMe targets. Run the code with the -h flag to see a list of available commands:
	
	$ python3 ./nvme_gw_cli.py -h
	usage: gwcli [-h] config_file {create_bdev,create_subsystem,create_namespace,allow_any_hosts,create_transport,create_listener} ...

	Manage NVMe gateways

	positional arguments:
	  config_file           Gateway configuration file
	  {create_bdev,create_subsystem,create_namespace,allow_any_hosts,create_transport,create_listener}

	optional arguments:
	  -h, --help            show this help message and exit

Specify a command with the -h flag to see more information about the command, including a list of optional and required parameters:

	$ python3 ./nvme_gw_cli.py nvme_gw.config create_bdev -h

> 	usage: gwcli config_file create_bdev [-h] -r RBD -p POOL [-n NAME] [-u USER] [-b BLOCK_SIZE]

Reviewer: It does not look like a good idea to make config_file as a positional mandatory argument. I am sure people would prefer not specify it on every invocation and rely on the cli finding it in a standard location (still having a possibility to override it with an option if needed).

Author: Agreed - we plan on having a default location that the config file can be read from if the arg is not specified.

Reviewer: Yes, then it should be an optional (e.g. `-f  config_file`), not positional, because you are not able to optionally specify a positional argument unless it is the last one. We can decide about a default location later, but why not make it not positional (i.e. specifying with `-f` option) from the start?

Author: > Yes, then it should be an optional (e.g. `-f config_file`), not positional, because you are not able to optionally specify a positional argument unless it is the last one. We can decide about a default location later, but why not make it not positional (i.e. specifying with `-f` option) from the start?

Agree. We'll change this.
